### PR TITLE
Allow pgAdmin to retrive master password from external script.#4769

### PR DESCRIPTION
**Pull request by yogeshmahajan-1903**

### CODE REVIEW

**Review by akshay-joshi — CHANGES_REQUESTED** [review]

File: web/config.py [hunk 1]
```diff
@@ -577,6 +577,22 @@
...
+# When pgAdmin stores a connection password,
+# it encrypts it using a key that is formed either from the master password, or
+# from the pgAdmin login password for the user.
+#
+# Below setting allows user to specify the path to a script or program
+# that will return a master password/key.
+# pgAdmin would use this encryption key derived from below script
+# This setting is used only in server mode. Also you can pass current
+# user name as argument to external script by specifying %u in config value
+# E.g. - MASTER_PASSWORD_HOOK = '<PATH>/passwdgen.sh %u'
```
[the remark this answers is not in the document]
Comment (yogeshmahajan-1903) [reply web/config.py]: Done

File: web/pgadmin/browser/__init__.py [hunk 2]
```diff
@@ -870,27 +875,51 @@ def set_master_password():
...
+                            'Please check that the hook utility is configured'
+                            ' correctly.'.format(config.MASTER_PASSWORD_HOOK))
+            return form_master_password_response(
+                existing=False,
+                present=False,
+                errmsg=error
+            )
+
+    # Master password is not applicable for Desktop mode and in server mode
+    # only when auth sources are oauth, kerberos, webserver.
```
[the remark this answers is not in the document]
Comment (yogeshmahajan-1903) [reply web/pgadmin/browser/__init__.py]: Done

File: web/pgadmin/browser/__init__.py [hunk 3]
```diff
@@ -906,16 +935,23 @@ def set_master_password():
             # master pass
             set_masterpass_check_text(data.get('password'))
 
+        # if password in request it empty, try to get it and if not
+        # retrieved check if masterpass_check is already set if yes
+        # provide user a pop to enter/get
+        # master password(present) from user but no option to reset(existing).
```
[the remark this answers is not in the document]
Comment (yogeshmahajan-1903) [reply web/pgadmin/browser/__init__.py]: Done

File: web/pgadmin/browser/__init__.py [hunk 4]
```diff
@@ -870,27 +875,51 @@ def set_master_password():
             )
             config.DISABLED_LOCAL_PASSWORD_STORAGE = True
 
-    # Master password is not applicable for server mode
+    # If master passowrd is specified try to retrieve master password
+    # & update data. If There is error while retrieving return with error
```
[the remark this answers is not in the document]
Comment (yogeshmahajan-1903) [reply web/pgadmin/browser/__init__.py]: Done

File: web/pgadmin/browser/__init__.py [hunk 5]
```diff
@@ -793,7 +793,12 @@ def reset_master_password():
                              KEY_RING_DESKTOP_USER.format(
                                  current_user.username), 'test')
     cleanup_master_password()
-    return make_json_response(data=get_crypt_key()[0])
+    status, crypt_key = get_crypt_key()
+    # Set masterpass_check as we need to input from user to set master
+    # password in case of MASTER_PASSWORD_HOOK
```
[the remark this answers is not in the document]
Comment (yogeshmahajan-1903) [reply web/pgadmin/browser/__init__.py]: Done

**Pull request merged**